<!DOCTYPE HTML>
<html lang="en" class="sidebar-visible no-js latte">
    <head>
        <!-- Book generated using mdBook -->
        <meta charset="UTF-8">
        <title>Drug resistance Prediction with Reference Graphs</title>
        <meta name="robots" content="noindex" />


        <!-- Custom HTML head -->
        
        <meta content="text/html; charset=utf-8" http-equiv="Content-Type">
        <meta name="description" content="">
        <meta name="viewport" content="width=device-width, initial-scale=1">
        <meta name="theme-color" content="#ffffff" />

        <link rel="icon" href="favicon.svg">
        <link rel="shortcut icon" href="favicon.png">
        <link rel="stylesheet" href="css/variables.css">
        <link rel="stylesheet" href="css/general.css">
        <link rel="stylesheet" href="css/chrome.css">
        <link rel="stylesheet" href="css/print.css" media="print">

        <!-- Fonts -->
        <link rel="stylesheet" href="FontAwesome/css/font-awesome.css">
        <link rel="stylesheet" href="fonts/fonts.css">

        <!-- Highlight.js Stylesheets -->
        <link rel="stylesheet" href="highlight.css">
        <link rel="stylesheet" href="tomorrow-night.css">
        <link rel="stylesheet" href="ayu-highlight.css">

        <!-- Custom theme stylesheets -->
        <link rel="stylesheet" href="./theme/catppuccin.css">
        <link rel="stylesheet" href="./theme/catppuccin-highlight.css">

    </head>
    <body>
        <!-- Provide site root to javascript -->
        <script type="text/javascript">
            var path_to_root = "";
            var default_theme = window.matchMedia("(prefers-color-scheme: dark)").matches ? "navy" : "latte";
        </script>

        <!-- Work around some values being stored in localStorage wrapped in quotes -->
        <script type="text/javascript">
            try {
                var theme = localStorage.getItem('mdbook-theme');
                var sidebar = localStorage.getItem('mdbook-sidebar');

                if (theme.startsWith('"') && theme.endsWith('"')) {
                    localStorage.setItem('mdbook-theme', theme.slice(1, theme.length - 1));
                }

                if (sidebar.startsWith('"') && sidebar.endsWith('"')) {
                    localStorage.setItem('mdbook-sidebar', sidebar.slice(1, sidebar.length - 1));
                }
            } catch (e) { }
        </script>

        <!-- Set the theme before any content is loaded, prevents flash -->
        <script type="text/javascript">
            var theme;
            try { theme = localStorage.getItem('mdbook-theme'); } catch(e) { }
            if (theme === null || theme === undefined) { theme = default_theme; }
            var html = document.querySelector('html');
            html.classList.remove('no-js')
            html.classList.remove('latte')
            html.classList.add(theme);
            html.classList.add('js');
        </script>

        <!-- Hide / unhide sidebar before it is displayed -->
        <script type="text/javascript">
            var html = document.querySelector('html');
            var sidebar = 'hidden';
            if (document.body.clientWidth >= 1080) {
                try { sidebar = localStorage.getItem('mdbook-sidebar'); } catch(e) { }
                sidebar = sidebar || 'visible';
            }
            html.classList.remove('sidebar-visible');
            html.classList.add("sidebar-" + sidebar);
        </script>

        <nav id="sidebar" class="sidebar" aria-label="Table of contents">
            <div class="sidebar-scrollbox">
                <ol class="chapter"><li class="chapter-item expanded affix "><a href="../../index.html">Introduction</a></li><li class="chapter-item expanded affix "><li class="part-title">Getting Started</li><li class="chapter-item expanded "><a href="installation.html"><strong aria-hidden="true">1.</strong> Installation</a></li><li class="chapter-item expanded affix "><li class="part-title">User Guide</li><li class="chapter-item expanded "><a href="guide/download.html"><strong aria-hidden="true">2.</strong> Download prebuilt index</a></li><li class="chapter-item expanded "><a href="guide/predict.html"><strong aria-hidden="true">3.</strong> Predict</a></li><li><ol class="section"><li class="chapter-item expanded "><a href="guide/predict-output.html"><strong aria-hidden="true">3.1.</strong> Prediction output</a></li></ol></li><li class="chapter-item expanded "><a href="guide/build/build.html"><strong aria-hidden="true">4.</strong> Build an index</a></li><li><ol class="section"><li class="chapter-item expanded "><a href="guide/build/vcf.html"><strong aria-hidden="true">4.1.</strong> VCF to build reference graph</a></li><li class="chapter-item expanded "><a href="guide/build/catalogue.html"><strong aria-hidden="true">4.2.</strong> Catalogue of mutations</a></li><li class="chapter-item expanded "><a href="guide/build/expert.html"><strong aria-hidden="true">4.3.</strong> Expert rules</a></li></ol></li></ol>
            </div>
            <div id="sidebar-resize-handle" class="sidebar-resize-handle"></div>
        </nav>

        <div id="page-wrapper" class="page-wrapper">

            <div class="page">
                                <div id="menu-bar-hover-placeholder"></div>
                <div id="menu-bar" class="menu-bar sticky bordered">
                    <div class="left-buttons">
                        <button id="sidebar-toggle" class="icon-button" type="button" title="Toggle Table of Contents" aria-label="Toggle Table of Contents" aria-controls="sidebar">
                            <i class="fa fa-bars"></i>
                        </button>
                        <button id="theme-toggle" class="icon-button" type="button" title="Change theme" aria-label="Change theme" aria-haspopup="true" aria-expanded="false" aria-controls="theme-list">
                            <i class="fa fa-paint-brush"></i>
                        </button>
                        <ul id="theme-list" class="theme-popup" aria-label="Themes" role="menu">
                            <li role="none"><button role="menuitem" class="theme" id="light">Light</button></li>
                            <li role="none"><button role="menuitem" class="theme" id="rust">Rust</button></li>
                            <li role="none"><button role="menuitem" class="theme" id="coal">Coal</button></li>
                            <li role="none"><button role="menuitem" class="theme" id="navy">Navy</button></li>
                            <li role="none"><button role="menuitem" class="theme" id="ayu">Ayu</button></li>
                            <li role="none"><button role="menuitem" class="theme" id="latte">Latte (default)</button></li>
                            <li role="none"><button role="menuitem" class="theme" id="frappe">Frapp√©</button></li>
                            <li role="none"><button role="menuitem" class="theme" id="macchiato">Macchiato</button></li>
                            <li role="none"><button role="menuitem" class="theme" id="mocha">Mocha</button></li>
                        </ul>
                        <button id="search-toggle" class="icon-button" type="button" title="Search. (Shortkey: s)" aria-label="Toggle Searchbar" aria-expanded="false" aria-keyshortcuts="S" aria-controls="searchbar">
                            <i class="fa fa-search"></i>
                        </button>
                    </div>

                    <h1 class="menu-title">Drug resistance Prediction with Reference Graphs</h1>

                    <div class="right-buttons">
                        <a href="print.html" title="Print this book" aria-label="Print this book">
                            <i id="print-button" class="fa fa-print"></i>
                        </a>
                        <a href="https://github.com/mbhall88/drprg" title="Git repository" aria-label="Git repository">
                            <i id="git-repository-button" class="fa fa-github"></i>
                        </a>

                    </div>
                </div>

                <div id="search-wrapper" class="hidden">
                    <form id="searchbar-outer" class="searchbar-outer">
                        <input type="search" id="searchbar" name="searchbar" placeholder="Search this book ..." aria-controls="searchresults-outer" aria-describedby="searchresults-header">
                    </form>
                    <div id="searchresults-outer" class="searchresults-outer hidden">
                        <div id="searchresults-header" class="searchresults-header"></div>
                        <ul id="searchresults">
                        </ul>
                    </div>
                </div>

                <!-- Apply ARIA attributes after the sidebar and the sidebar toggle button are added to the DOM -->
                <script type="text/javascript">
                    document.getElementById('sidebar-toggle').setAttribute('aria-expanded', sidebar === 'visible');
                    document.getElementById('sidebar').setAttribute('aria-hidden', sidebar !== 'visible');
                    Array.from(document.querySelectorAll('#sidebar a')).forEach(function(link) {
                        link.setAttribute('tabIndex', sidebar === 'visible' ? 0 : -1);
                    });
                </script>

                <div id="content" class="content">
                    <main>
                        <p>üë©‚Äç‚öïDr. PRG - <strong>D</strong>rug <strong>r</strong>esistance <strong>P</strong>rediction with <strong>R</strong>eference <strong>G</strong>raphsÔ∏èüë®‚Äç‚öïÔ∏è</p>
<p><a href="https://codecov.io/gh/mbhall88/drprg"><img src="https://codecov.io/gh/mbhall88/drprg/branch/main/graph/badge.svg?token=2WAA6MZRKK" alt="codecov" /></a>
<a href="https://github.com/mbhall88/drprg/actions/workflows/rust.yml"><img src="https://github.com/mbhall88/drprg/actions/workflows/rust.yml/badge.svg?branch=main" alt="Rust" /></a></p>
<p>Full documentation: <a href="https://mbh.sh/drprg/">https://mbh.sh/drprg/</a></p>
<p>As the name suggests, DrPRG is a tool for predicting drug resistance from sequencing
data. It can be used for any species, provided an index is available for that species.
The <a href="https://mbh.sh/drprg/">documentation</a> outlines which species have prebuilt indices and also a guide
for how to create your own.</p>
<h2 id="quick-usage"><a class="header" href="#quick-usage">Quick usage</a></h2>
<p>Download the latest <em>M. tuberculosis</em> prebuilt index</p>
<pre><code>drprg index --download mtb
</code></pre>
<p>Predict resistance from an Illumina fastq</p>
<pre><code>drprg predict -x mtb -i reads.fq --illumina -o outdir/
</code></pre>
<h3 id="help"><a class="header" href="#help">Help</a></h3>
<pre><code>$ drprg -h
Drug Resistance Prediction with Reference Graphs

Usage: drprg [OPTIONS] &lt;COMMAND&gt;

Commands:
  build    Build an index to predict resistance from
  predict  Predict drug resistance
  index    Download and interact with indices
  help     Print this message or the help of the given subcommand(s)

Options:
  -v, --verbose        Use verbose output
  -t, --threads &lt;INT&gt;  Maximum number of threads to use [default: 1]
  -h, --help           Print help (see more with '--help')
  -V, --version        Print version
</code></pre>
<h2 id="citation"><a class="header" href="#citation">Citation</a></h2>
<p>Preprint coming soon...</p>
<div style="break-before: page; page-break-before: always;"></div><h1 id="installation"><a class="header" href="#installation">Installation</a></h1>
<h2 id="conda"><a class="header" href="#conda">Conda</a></h2>
<pre><code>conda install drprg
</code></pre>
<h2 id="container"><a class="header" href="#container">Container</a></h2>
<p><a href="https://quay.io/repository/mbhall88/drprg"><img src="https://quay.io/repository/mbhall88/drprg/status" alt="Docker Repository on Quay" title="Docker Repository on Quay" /></a></p>
<p>A Docker container is available for all commits/branches/versions. To view the available
tags, visit <a href="https://quay.io/repository/mbhall88/drprg?tab=tags">https://quay.io/repository/mbhall88/drprg?tab=tags</a></p>
<p>For example, to use the latest commit on the <code>main</code> branch, the URI is</p>
<pre><code>$ TAG=&quot;latest&quot;
$ URI=&quot;quay.io/mbhall88/drprg:$TAG&quot;
</code></pre>
<h3 id="docker"><a class="header" href="#docker">Docker</a></h3>
<p>To run <code>drprg</code> using the above container with Docker</p>
<pre><code>$ docker pull &quot;$URI&quot;
$ docker run -it &quot;$URI&quot; drprg --help
</code></pre>
<h3 id="singularity"><a class="header" href="#singularity">Singularity</a></h3>
<p>To run <code>drprg</code> using the above container with <a href="https://sylabs.io/">Singularity</a></p>
<pre><code>$ singularity exec &quot;docker://$URI&quot; drprg --help
</code></pre>
<h2 id="local"><a class="header" href="#local">Local</a></h2>
<p><strong>Minimum supported Rust version</strong>: <code>1.65.0</code></p>
<pre><code>$ cargo build --release
$ target/release/drprg -h
</code></pre>
<h3 id="dependencies"><a class="header" href="#dependencies">Dependencies</a></h3>
<p><code>drprg</code> relies on:</p>
<ul>
<li><a href="https://github.com/rmcolq/pandora"><code>pandora</code></a></li>
<li><a href="https://samtools.github.io/bcftools/bcftools.html"><code>bcftools</code></a></li>
<li><a href="https://github.com/leoisl/make_prg/"><code>make_prg</code></a></li>
<li><a href="https://mafft.cbrc.jp/alignment/software/"><code>mafft</code></a></li>
</ul>
<p>You can install the dependencies using the provided <a href="https://github.com/casey/just"><code>justfile</code></a></p>
<pre><code class="language-shell script"># all dependencies
$ just deps
# pandora only
$ just pandora
# make_prg only
$ just makeprg
# mafft only
$ just mafft
# bcftools only
$ just bcftools
</code></pre>
<p>By default, the external dependencies will be downloaded to <code>src/ext</code>. This can be
changed by specifying a path to <code>EXTDIR</code> when installing the external dependencies.</p>
<pre><code class="language-shell script">$ just deps EXTDIR=&quot;some/other/dir&quot;
</code></pre>
<div style="break-before: page; page-break-before: always;"></div><h1 id="download-prebuilt-index"><a class="header" href="#download-prebuilt-index">Download prebuilt index</a></h1>
<p>The <code>index</code> subcommand can be used to download or list available indices.</p>
<p>To download the latest index version for all species</p>
<pre><code>drprg index --download
# is the same as
drprg index --download all
</code></pre>
<p>If you only want the latest for a particular species</p>
<pre><code>drprg index --download mtb
# is the same as
drprg index --download mtb@latest
</code></pre>
<p>If you want to see what indices are available in the first place</p>
<pre><code>drprg index --list
+--------------+---------+----------+------------+
| Name         | Species | Version  | Downloaded |
+--------------+---------+----------+------------+
| mtb@20230308 | mtb     | 20230308 | N          |
+--------------+---------+----------+------------+
</code></pre>
<p>You can specify a version with the following syntax</p>
<pre><code>drprg index --download mtb@&lt;version&gt;
</code></pre>
<p>where <code>&lt;version&gt;</code> is the version you want. If no version is provided, <code>latest</code> is used.</p>
<h2 id="output-directory"><a class="header" href="#output-directory">Output directory</a></h2>
<p>By default, indices are stored in <code>$HOME/.drprg/</code>, nested as <code>species/species-version</code>.
So downloading version <code>20230308</code> of species <code>mtb</code> will produce an index
at <code>$HOME/.drprg/mtb/mtb-20230308/</code>.</p>
<p>You can change the default output directory (<code>$HOME/.drprg/</code>) with the <code>--outdir</code>
option. If you do this, you'll need to pass the full path to the index when
using <a href="guide/predict.html"><code>predict</code></a>.</p>
<h2 id="full-usage"><a class="header" href="#full-usage">Full usage</a></h2>
<pre><code>$ drprg index --help
Download and interact with indices

Usage: drprg index [OPTIONS] [NAME]

Arguments:
  [NAME]
          The name/path of the index to download

          [default: all]

Options:
  -d, --download
          Download a prebuilt index

  -v, --verbose
          Use verbose output

  -l, --list
          List all available (and downloaded) indices

  -t, --threads &lt;INT&gt;
          Maximum number of threads to use

          Use 0 to select the number automatically

          [default: 1]

  -o, --outdir &lt;DIR&gt;
          Index directory

          Use this if your indices are not in a default location, or you want to download them to a non-default location

          [default: $HOME/.drprg/]

  -F, --force
          Overwrite any existing indices

  -h, --help
          Print help (see a summary with '-h')
</code></pre>
<div style="break-before: page; page-break-before: always;"></div><h1 id="predict"><a class="header" href="#predict">Predict</a></h1>
<p>The <code>predict</code> subcommand is used to predict resistance for a sample from an index.</p>
<p>At its simplest</p>
<pre><code>drprg predict -i reads.fq -x mtb -o outdir
</code></pre>
<p><code>drprg</code> is a bit &quot;new-age&quot; in that it assumes the reads are Nanopore. If they're
Illumina, use the <code>-I/--illumina</code> option.</p>
<p>See <a href="guide/./predict-output.html">Prediction Output</a> documentation for a detailed description of
what results/output files and formats to expect.</p>
<h2 id="required"><a class="header" href="#required">Required</a></h2>
<h3 id="index"><a class="header" href="#index">Index</a></h3>
<p>The index is provided via the <code>-x/--index</code> option. It can either be a path to an index,
or the name of a <a href="guide/./download.html">downloaded index</a>. As with the <a href="guide/./download.html"><code>index</code></a>
subcommmand, you can specify a version if you don't want to use the latest.</p>
<h3 id="input-reads"><a class="header" href="#input-reads">Input reads</a></h3>
<p>A fastq (or fasta) file of the reads you want to predict resistance from - provided via
the <code>-i/--input</code> option. If you have paired reads in two files, simply combine them and
pass the combined file - interleave order doesn't matter. For example</p>
<pre><code>cat r1.fq r2.fq &gt; combined.fq
drprg predict -i combined.fq ...
</code></pre>
<p><code>gzip</code>-compressed files are also accepted.</p>
<h2 id="optional"><a class="header" href="#optional">Optional</a></h2>
<h3 id="sample-name"><a class="header" href="#sample-name">Sample name</a></h3>
<p>Identifier to use for your output files. By default, it will be set to the file name
prefix (e.g. <code>name</code> for a fastq named <code>name.fq.gz</code>). Provided via the <code>-s/--sample</code>
option.</p>
<h3 id="minimum-allele-frequency"><a class="header" href="#minimum-allele-frequency">Minimum allele frequency</a></h3>
<p>Provided via the <code>-f/--maf</code> option. If an alternate allele has at least this fraction of
the depth, a minor resistance (&quot;r&quot;) prediction is made. By default, this is set to <code>1.0</code>
for Nanopore data (i.e. minor allele detection is off) and <code>0.1</code> when using
the <code>--illumina</code> option. For example, if a variant is called as the reference allele for
Illumina reads, but an alternate allele has more than 10% of the depth on that position,
a minor resistance call is made for the alternate allele.</p>
<h3 id="ignore-synonymous"><a class="header" href="#ignore-synonymous">Ignore synonymous</a></h3>
<p>Using the <code>-S/--ignore-synonymous</code> option will prevent synonymous mutations from
appearing as unknown resistance calls. However, any synonymous mutations in the
catalogue will still be considered.</p>
<h2 id="quick-usage-1"><a class="header" href="#quick-usage-1">Quick usage</a></h2>
<pre><code>$ drprg predict -h
Predict drug resistance

Usage: drprg predict [OPTIONS] --index &lt;DIR&gt; --input &lt;FILE&gt;

Options:
  -v, --verbose        Use verbose output
  -t, --threads &lt;INT&gt;  Maximum number of threads to use [default: 1]
  -h, --help           Print help (see more with '--help')

Input/Output:
  -x, --index &lt;DIR&gt;      Name of a downloaded index or path to an index
  -i, --input &lt;FILE&gt;     Reads to predict resistance from
  -o, --outdir &lt;DIR&gt;     Directory to place output [default: .]
  -s, --sample &lt;SAMPLE&gt;  Identifier to use for the sample
  -I, --illumina         Sample reads are from Illumina sequencing

Filter:
  -S, --ignore-synonymous     Ignore unknown (off-catalogue) variants that cause a synonymous substitution
  -f, --maf &lt;FLOAT[0.0-1.0]&gt;  Minimum allele frequency to call variants [default: 1]
</code></pre>
<h2 id="full-usage-1"><a class="header" href="#full-usage-1">Full usage</a></h2>
<pre><code>$ drprg predict --help
Predict drug resistance

Usage: drprg predict [OPTIONS] --index &lt;DIR&gt; --input &lt;FILE&gt;

Options:
  -p, --pandora &lt;FILE&gt;
          Path to pandora executable. Will try in src/ext or $PATH if not given

  -v, --verbose
          Use verbose output

  -m, --makeprg &lt;FILE&gt;
          Path to make_prg executable. Will try in src/ext or $PATH if not given

  -t, --threads &lt;INT&gt;
          Maximum number of threads to use

          Use 0 to select the number automatically

          [default: 1]

  -M, --mafft &lt;FILE&gt;
          Path to MAFFT executable. Will try in src/ext or $PATH if not given

  -h, --help
          Print help (see a summary with '-h')

Input/Output:
  -x, --index &lt;DIR&gt;
          Name of a downloaded index or path to an index

  -i, --input &lt;FILE&gt;
          Reads to predict resistance from

          Both fasta and fastq are accepted, along with compressed or uncompressed.

  -o, --outdir &lt;DIR&gt;
          Directory to place output

          [default: .]

  -s, --sample &lt;SAMPLE&gt;
          Identifier to use for the sample

          If not provided, this will be set to the input reads file path prefix

  -I, --illumina
          Sample reads are from Illumina sequencing

Filter:
  -S, --ignore-synonymous
          Ignore unknown (off-catalogue) variants that cause a synonymous substitution

  -f, --maf &lt;FLOAT[0.0-1.0]&gt;
          Minimum allele frequency to call variants

          If an alternate allele has at least this fraction of the depth, a minor resistance (&quot;r&quot;) prediction is made. Set to 1 to disable. If --illumina is passed, the default is 0.1

          [default: 1]

      --debug
          Output debugging files. Mostly for development purposes

  -d, --min-covg &lt;INT&gt;
          Minimum depth of coverage allowed on variants

          [default: 3]

  -D, --max-covg &lt;INT&gt;
          Maximum depth of coverage allowed on variants

          [default: 2147483647]

  -b, --min-strand-bias &lt;FLOAT&gt;
          Minimum strand bias ratio allowed on variants

          For example, setting to 0.25 requires &gt;=25% of total (allele) coverage on both strands for an allele.

          [default: 0.01]

  -g, --min-gt-conf &lt;FLOAT&gt;
          Minimum genotype confidence (GT_CONF) score allow on variants

          [default: 0]

  -L, --max-indel &lt;INT&gt;
          Maximum (absolute) length of insertions/deletions allowed

  -K, --min-frs &lt;FLOAT&gt;
          Minimum fraction of read support

          For example, setting to 0.9 requires &gt;=90% of coverage for the variant to be on the called allele

          [default: 0]
</code></pre>
<div style="break-before: page; page-break-before: always;"></div><h1 id="prediction-output"><a class="header" href="#prediction-output">Prediction Output</a></h1>
<p>Inside the <code>predict</code> output directory (<code>-o/--outdir</code>) you'll find a collection of files
and directories. We outline the files that are likely to be of interest to most users.</p>
<h2 id="prediction-json"><a class="header" href="#prediction-json">Prediction JSON</a></h2>
<p>This is <code>&lt;sample&gt;.drprg.json</code>. The value given to the <code>-s/--sample</code>
option dictates the name - i.e. <code>&lt;sample&gt;.drprg.json</code>.</p>
<p>This is <strong>the only file most users will want/need to interact with</strong>. It contains the
resistance prediction for each drug in the index's catalogue.</p>
<h4 id="example"><a class="header" href="#example">Example</a></h4>
<p>This is a trimmmed (toy) example JSON output for a sample</p>
<pre><code class="language-json">{
  &quot;genes&quot;: {
    &quot;absent&quot;: [
      &quot;ahpC&quot;
    ],
    &quot;present&quot;: [
      &quot;embA&quot;,
      &quot;embB&quot;,
      &quot;ethA&quot;,
      &quot;fabG1&quot;,
      &quot;gid&quot;,
      &quot;gyrA&quot;,
      &quot;gyrB&quot;,
      &quot;inhA&quot;,
      &quot;katG&quot;,
      &quot;pncA&quot;,
      &quot;rpoB&quot;,
      &quot;rrs&quot;
    ]
  },
  &quot;sample&quot;: &quot;toy&quot;,
  &quot;susceptibility&quot;: {
    &quot;Amikacin&quot;: {
      &quot;evidence&quot;: [
        {
          &quot;gene&quot;: &quot;rrs&quot;,
          &quot;residue&quot;: &quot;DNA&quot;,
          &quot;variant&quot;: &quot;A1401X&quot;,
          &quot;vcfid&quot;: &quot;b815ed3f&quot;
        }
      ],
      &quot;predict&quot;: &quot;F&quot;
    },
    &quot;Ethambutol&quot;: {
      &quot;evidence&quot;: [
        {
          &quot;gene&quot;: &quot;embB&quot;,
          &quot;residue&quot;: &quot;PROT&quot;,
          &quot;variant&quot;: &quot;M306I&quot;,
          &quot;vcfid&quot;: &quot;a290b118&quot;
        }
      ],
      &quot;predict&quot;: &quot;r&quot;
    },
    &quot;Ethionamide&quot;: {
      &quot;evidence&quot;: [
        {
          &quot;gene&quot;: &quot;ethA&quot;,
          &quot;residue&quot;: &quot;PROT&quot;,
          &quot;variant&quot;: &quot;A381P&quot;,
          &quot;vcfid&quot;: &quot;169f75d4&quot;
        }
      ],
      &quot;predict&quot;: &quot;U&quot;
    },
    &quot;Isoniazid&quot;: {
      &quot;evidence&quot;: [
        {
          &quot;gene&quot;: &quot;fabG1&quot;,
          &quot;residue&quot;: &quot;DNA&quot;,
          &quot;variant&quot;: &quot;G-17T&quot;,
          &quot;vcfid&quot;: &quot;de9b689e&quot;
        },
        {
          &quot;gene&quot;: &quot;katG&quot;,
          &quot;residue&quot;: &quot;PROT&quot;,
          &quot;variant&quot;: &quot;S315T&quot;,
          &quot;vcfid&quot;: &quot;acaa8ca2&quot;
        }
      ],
      &quot;predict&quot;: &quot;R&quot;
    },
    &quot;Levofloxacin&quot;: {
      &quot;evidence&quot;: [],
      &quot;predict&quot;: &quot;S&quot;
    }
  },
  &quot;version&quot;: {
    &quot;drprg&quot;: &quot;0.1.0&quot;,
    &quot;index&quot;: &quot;20230308&quot;
  }
}
</code></pre>
<p>The keys of the JSON are</p>
<ul>
<li><code>genes</code>: This contains a list of genes in the index reference graph which are present
and absent</li>
<li><code>sample</code>: The value passed to the <code>-s/--sample</code> option</li>
<li><code>susceptibility</code>: The keys of this entry are the drugs in the index catalogue. Each
drug's entry contains <a href="guide/predict-output.html#evidence"><code>evidence</code></a> supporting the value in
the <a href="guide/predict-output.html#predict"><code>predict</code></a> section.</li>
</ul>
<h3 id="predict-1"><a class="header" href="#predict-1">Predict</a></h3>
<p>The <code>predict</code> entry for a drug is the resistance prediction for the sample. Possible
values are</p>
<ul>
<li><code>S</code>: susceptible. This is the &quot;default&quot; prediction. If no mutations are detected for
the sample, it is assumed to be susceptible</li>
<li><code>F</code>: failed. Genotyping failed for one or more mutations for this drug. See
the <a href="guide/predict-output.html#prediction-vcf">prediction VCF</a> for more information</li>
<li><code>U</code>: unknown. One or more mutations that are not present in the index catalogue were
detected in a gene associated with this drug</li>
<li><code>R</code>: resistant. One or more mutations from the index catalogue that confer resistance
were detected</li>
<li><code>u</code> or <code>r</code>: The same as the uppercase versions, but the mutation(s) were detected in
a <a href="guide/./predict.html#minimum-allele-frequency">minor allele</a>.</li>
</ul>
<h3 id="evidence"><a class="header" href="#evidence">Evidence</a></h3>
<p>This is a list of the mutations supporting the prediction. The <code>residue</code> is one of <code>DNA</code>
or <code>PROT</code> indicating whether the mutation describes a nucleotide or amino acid change,
respectively.</p>
<p>The <code>variant</code> is of the form <code>&lt;ref&gt;&lt;pos&gt;&lt;alt&gt;</code>; where <code>&lt;ref&gt;</code> is the reference sequence
at position <code>&lt;pos&gt;</code> and <code>&lt;alt&gt;</code> is the nucleotide/amino acid the reference is changed
to. See the <a href="guide/./build/catalogue.html">catalogue docs</a> for more information.</p>
<p>The <code>vcfid</code> is the value in <a href="guide/predict-output.html#prediction-vcf">the VCF</a> <code>ID</code> column for this mutation,
making it easier to find a mutation in <a href="guide/predict-output.html#prediction-vcf">the VCF</a>.</p>
<h2 id="prediction-vcf"><a class="header" href="#prediction-vcf">Prediction VCF</a></h2>
<p>This is <code>&lt;sample&gt;.drprg.bcf</code>. As this file is a <a href="https://samtools.github.io/hts-specs/VCFv4.4.pdf">BCF</a>, you will need to
use <a href="https://github.com/samtools/bcftools/"><code>bcftools</code></a> to view it - e.g. <code>bcftools view sample.drprg.bcf</code>.</p>
<p>You should only need to interact with this file if you want further information about
the exact evidence supporting a mutation being called, why a mutation was called as
failed (<code>F</code>), or why a mutation <em>wasn't</em> called. For those mutations in the JSON, you
can easily look them up using the <code>vcfid</code>, which can be found in the <code>ID</code> (third)
column of the BCF. Or you can just use <a href="https://www.man7.org/linux/man-pages/man1/grep.1.html"><code>grep</code></a>. For example, to look up the
Isoniazid S315T mutation
in <em>katG</em> from <a href="guide/predict-output.html#example">the example</a> you can use</p>
<pre><code>$ bcftools view sample.drprg.bcf | grep acaa8ca2
katG    1044    acaa8ca2        GC      AC,CA,CC        .       PASS    VC=PH_SNPs;GRAPHTYPE=SIMPLE;PDP=0,0.0123457,0,0.987654;VARID=katG_S315T;PREDICT=R       GT:MEAN_FWD_COVG:MEAN_REV_COVG:MED_FWD_COVG:MED_REV_COVG:SUM_FWD_COVG:SUM_REV_COVG:GAPS:LIKELIHOOD:GT_CONF      3:0,1,0,42:0,0,0,38:0,1,0,42:0,0,0,38:0,1,0,127:0,0,0,116:1,1,1,0:-523.019,-514.096,-523.019,-7.87925:506.217
</code></pre>
<p>All <code>INFO</code> and <code>FORMAT</code> fields are defined in the header of the BCF file. We recommend
reading the <a href="https://samtools.github.io/hts-specs/VCFv4.4.pdf">VCF/BCF</a> specifications for help with how to interpret data in a VCF
file.</p>
<div style="break-before: page; page-break-before: always;"></div><h1 id="build-an-index"><a class="header" href="#build-an-index">Build an index</a></h1>
<p>The <code>build</code> subcommand is used to build a custom index to <a href="guide/build/../predict.html"><code>predict</code></a>
from.</p>
<pre><code>$ drprg build -a annotation.gff3 -i catalogue.tsv -f ref.fa -o outdir
</code></pre>
<h2 id="required-1"><a class="header" href="#required-1">Required</a></h2>
<h3 id="catalogue"><a class="header" href="#catalogue">Catalogue</a></h3>
<p>Also referred to as a &quot;panel&quot; sometimes. This is the catalogue of mutations that confer
resistance (or susceptibility). See <a href="guide/build/./catalogue.html">the Catalogue docs</a> for more
details of what format this file must take. Provided via the <code>-i/--panel</code> option.</p>
<h3 id="reference-genome"><a class="header" href="#reference-genome">Reference genome</a></h3>
<p>A FASTA file of the reference genome that will act as the &quot;skeleton&quot; for the reference
graph. Provided via the <code>-f/--fasta</code> option.</p>
<h3 id="annotation"><a class="header" href="#annotation">Annotation</a></h3>
<p>The annotation is a <a href="https://github.com/The-Sequence-Ontology/Specifications/blob/master/gff3.md">GFF3</a> file for the species you are building the index for.
Provided via the <code>-a/--gff</code> option.</p>
<h3 id="vcf"><a class="header" href="#vcf">VCF</a></h3>
<p>A VCF to build the reference graph from. See <a href="guide/build/./vcf.html">VCF</a> for a detailed description
of this file. Provided via the <code>-b/--vcf</code> option. This option is mutually exclusive
with <a href="guide/build/build.html#prebuilt-reference-graph"><code>--prebuilt-prg</code></a>.</p>
<h2 id="optional-1"><a class="header" href="#optional-1">Optional</a></h2>
<h3 id="expert-rules"><a class="header" href="#expert-rules">Expert rules</a></h3>
<p>These describe a set of rules - rather than a single mutation - that cause resistance.
See <a href="guide/build/./expert.html">Expert rules</a> for a detailed description of the rules. Provided via
the <code>-r/--rules</code> option.</p>
<h3 id="padding"><a class="header" href="#padding">Padding</a></h3>
<p>Number of bases to add to the start and end of each gene (locus) in the reference graph.
Set to 100 by default. Provided via the <code>-P/--padding</code> option.</p>
<h3 id="prebuilt-reference-graph"><a class="header" href="#prebuilt-reference-graph">Prebuilt reference graph</a></h3>
<p><strong>Advanced users only.</strong> For those who know how to build their own reference graphs, and
would prefer to do that themselves by hand, rather than let DrPRG do it for them. The
value passed to this option must be a directory containing:</p>
<ul>
<li>A reference graph file named <code>dr.prg</code></li>
<li>A directory of multiple sequence alignments (<code>/msas/</code>) that the reference graph was
built from. There must be a fasta file alignment for each gene in this
directory <code>&lt;gene&gt;.fa</code>.</li>
<li>An optional <a href="https://github.com/rmcolq/pandora"><code>pandora</code></a> index. If not present, one will be created.</li>
</ul>
<p>The reference graph is expected to contain the reference sequence (
with <a href="guide/build/build.html#padding">padding</a>) for each gene according to the <a href="guide/build/build.html#annotation">annotation</a>
and <a href="guide/build/build.html#reference-genome">reference sequence</a> provided.</p>
<p>Provided with the <code>-d/--prebuilt-prg</code> option.</p>
<h3 id="version"><a class="header" href="#version">Version</a></h3>
<p>The version name for the index. Set to the current date (<code>YYYYMMDD</code>) by default.
Provided via the <code>--version</code> option.</p>
<h2 id="quick-usage-2"><a class="header" href="#quick-usage-2">Quick usage</a></h2>
<pre><code>$ drprg build -h
Build an index to predict resistance from

Usage: drprg build [OPTIONS] --gff &lt;FILE&gt; --panel &lt;FILE&gt; --fasta &lt;FILE&gt;

Options:
  -v, --verbose            Use verbose output
  -t, --threads &lt;INT&gt;      Maximum number of threads to use [default: 1]
  -P, --padding &lt;INT&gt;      Number of bases of padding to add to start and end of each gene [default: 100]
  -I, --no-fai             Don't index --fasta if an index doesn't exist
  -C, --no-csi             Don't index --vcf if an index doesn't exist
      --version &lt;VERSION&gt;  Version to use for the index [default: 20230405]
  -h, --help               Print help (see more with '--help')

Input/Output:
  -a, --gff &lt;FILE&gt;          Annotation file that will be used to gather information about genes in catalogue
  -i, --panel &lt;FILE&gt;        Panel/catalogue to build index for
  -f, --fasta &lt;FILE&gt;        Reference genome in FASTA format (must be indexed with samtools faidx)
  -b, --vcf &lt;FILE&gt;          An indexed VCF to build the index PRG from. If not provided, then a prebuilt PRG must be given. See `--prebuilt-prg`
  -o, --outdir &lt;DIR&gt;        Directory to place output [default: .]
  -d, --prebuilt-prg &lt;DIR&gt;  A prebuilt PRG to use
  -r, --rules &lt;FILE&gt;        &quot;Expert rules&quot; to be applied in addition to the catalogue
</code></pre>
<h2 id="full-usage-2"><a class="header" href="#full-usage-2">Full usage</a></h2>
<pre><code>$ drprg build --help
Build an index to predict resistance from

Usage: drprg build [OPTIONS] --gff &lt;FILE&gt; --panel &lt;FILE&gt; --fasta &lt;FILE&gt;

Options:
  -p, --pandora &lt;FILE&gt;
          Path to pandora executable. Will try in src/ext or $PATH if not given

  -v, --verbose
          Use verbose output

  -m, --makeprg &lt;FILE&gt;
          Path to make_prg executable. Will try in src/ext or $PATH if not given

  -t, --threads &lt;INT&gt;
          Maximum number of threads to use

          Use 0 to select the number automatically

          [default: 1]

  -M, --mafft &lt;FILE&gt;
          Path to MAFFT executable. Will try in src/ext or $PATH if not given

  -B, --bcftools &lt;FILE&gt;
          Path to bcftools executable. Will try in src/ext or $PATH if not given

  -P, --padding &lt;INT&gt;
          Number of bases of padding to add to start and end of each gene

          [default: 100]

  -l, --match-len &lt;INT&gt;
          Minimum number of consecutive characters which must be identical for a match in make_prg

          [default: 5]

  -N, --max-nesting &lt;INT&gt;
          Maximum nesting level when constructing the reference graph with make_prg

          [default: 5]

  -k, --pandora-k &lt;INT&gt;
          Kmer size to use for pandora

          [default: 15]

  -w, --pandora-w &lt;INT&gt;
          Window size to use for pandora

          [default: 11]

  -I, --no-fai
          Don't index --fasta if an index doesn't exist

  -C, --no-csi
          Don't index --vcf if an index doesn't exist

      --version &lt;VERSION&gt;
          Version to use for the index

          [default: 20230405]

  -h, --help
          Print help (see a summary with '-h')

Input/Output:
  -a, --gff &lt;FILE&gt;
          Annotation file that will be used to gather information about genes in catalogue

  -i, --panel &lt;FILE&gt;
          Panel/catalogue to build index for

  -f, --fasta &lt;FILE&gt;
          Reference genome in FASTA format (must be indexed with samtools faidx)

  -b, --vcf &lt;FILE&gt;
          An indexed VCF to build the index PRG from. If not provided, then a prebuilt PRG must be given. See `--prebuilt-prg`

  -o, --outdir &lt;DIR&gt;
          Directory to place output

          [default: .]

  -d, --prebuilt-prg &lt;DIR&gt;
          A prebuilt PRG to use.

          Only build the panel VCF and reference sequences - not the PRG. This directory MUST contain a PRG file named `dr.prg`, along, with a directory called `msas/` that contains an MSA fasta file for each gene `&lt;gene&gt;.fa`. There can optionally also be a pandora index file, but if not, the indexing will be performed by drprg. Note: the PRG is expected to contain the reference sequence for each gene according to the annotation and reference genome given (along with padding) and must be in the forward strand orientation.

  -r, --rules &lt;FILE&gt;
          &quot;Expert rules&quot; to be applied in addition to the catalogue.

          CSV file with blanket rules that describe resistance (or susceptibility). The columns are &lt;variant type&gt;,&lt;gene&gt;,&lt;start&gt;,&lt;end&gt;,&lt;drug(s)&gt;. See the docs for a detailed explanation.
</code></pre>
<div style="break-before: page; page-break-before: always;"></div><h1 id="vcf-to-build-reference-graph"><a class="header" href="#vcf-to-build-reference-graph">VCF to build reference graph</a></h1>
<p>The input VCF (<code>-b/--vcf</code>) for <code>build</code> has some very particular requirements. If you
aren't very familiar with the <a href="https://samtools.github.io/hts-specs/VCFv4.3.pdf">formal VCF specifications</a>, please read them
first.</p>
<p>This VCF file is what <code>drprg</code> uses to construct the reference graph from. Variants from
it are applied to the <a href="guide/build/./build.html#reference-genome">skeleton reference genome</a>. It
represents the population variation you would like to build into the reference graph.
For a greater understanding of these concepts, we recommend
reading <a href="https://github.com/rmcolq/pandora">this paper</a>.</p>
<p>To add variants from multiple samples, simply use a multi-sample VCF.</p>
<p>The <code>CHROM</code> field should be the gene name. As such, the <code>POS</code> is with respect to the
gene. It's important to ensure the <code>POS</code> takes into account
the <a href="guide/build/./build.html#padding">padding</a> that will be used when running <code>build</code>. A helper
script for converting a VCF from reference genome coordinates to gene coordinates can be
found <a href="https://github.com/mbhall88/drprg/blob/main/scripts/extract_panel_genes_from_vcf.py">here</a>. In addition, <code>drprg</code> will raise an error if the coordinates
are incorrect with respect to the annotation and reference genome you provide.</p>
<div style="break-before: page; page-break-before: always;"></div><h1 id="catalogue-of-mutations"><a class="header" href="#catalogue-of-mutations">Catalogue of mutations</a></h1>
<p>Also referred to as a &quot;panel&quot; sometimes. This is the catalogue of mutations that confer
resistance (or susceptibility). Provided via the <code>-i/--panel</code> option.</p>
<p>This file is a tab-delimited (TSV) file with four columns</p>
<ol>
<li>The gene name of the mutation</li>
<li>The mutation in the form <code>&lt;ref&gt;&lt;pos&gt;&lt;alt&gt;</code>, where <code>&lt;ref&gt;</code> is the reference nucleotide
or amino acid, <code>&lt;pos&gt;</code> is the 1-based position in the gene/protein, and <code>&lt;alt&gt;</code> is
the nucleotide or amino acid the reference is changed to. <code>&lt;pos&gt;</code> is with respect to
the type of mutation - i.e. if the mutation is a amino acid change, the <code>&lt;pos&gt;</code> must
be position within the protein (codon position). Promoter mutations can be specied
with a negative symbol - e.g. -10 means 10 nucleotide before the first position in
the gene. One-letter codes are to be used for amino acids.</li>
<li>The residue type - <code>DNA</code> (nucleotide change) or <code>PROT</code> (amino acid change).</li>
<li>Comma-separated list of the drug(s) the mutation confers resistance to. Use <code>NONE</code> if
the mutation is not associated with resistance. </li>
</ol>
<h2 id="example-1"><a class="header" href="#example-1">Example</a></h2>
<pre><code>pncA    TCG196TAG   DNA Pyrazinamide
pncA    T142R   PROT    Pyrazinamide
tlyA    S159*   PROT    Capreomycin
embB    M306N   PROT    Ethambutol
rpoB    C1275CCA    DNA Rifampicin
gid R137P   PROT    Streptomycin
gid R118S   PROT    Streptomycin
tlyA    G123GC  DNA Capreomycin
pncA    G97D    PROT    Pyrazinamide
fabG1   C-15X   DNA Ethionamide,Isoniazid
</code></pre>
<div style="break-before: page; page-break-before: always;"></div><h1 id="expert-rules-1"><a class="header" href="#expert-rules-1">Expert rules</a></h1>
<p>These are blanket rules that describe resistance (or susceptibility). The file is a CSV
with each row representing a rule and is passed to <code>drprg build</code> via the <code>--rules</code>
option. The format of each row is</p>
<pre><code class="language-csv">vartype,gene,start,end,drug
</code></pre>
<ol>
<li><code>vartype</code>: the variant type of the rule. Supported types are:
<ul>
<li><a href="https://www.genome.gov/genetics-glossary/Frameshift-Mutation"><code>frameshift</code></a> - Any insertion or deletion whose length is not a
multiple of three</li>
<li><a href="https://www.genome.gov/genetics-glossary/Missense-Mutation"><code>missense</code></a> - A DNA change that results in a different amino acid</li>
<li><a href="https://www.genome.gov/genetics-glossary/Nonsense-Mutation"><code>nonsense</code></a> - A DNA change that results in a stop codon instead of an
amino acid</li>
<li><code>absence</code> - Gene is absent</li>
</ul>
</li>
<li><code>gene</code>: the name of the gene the rule applies to</li>
<li><code>start</code>: An optional start position for the rule to apply from. The position is in
codon coordinates where the rule applies to amino acid changes and is 1-based
inclusive. If not provided, the start of the gene is inferred. If you want to include
the upstream (promoter) region of the gene, use negative coordinates.</li>
<li><code>end</code>: An optional end position for the rule to apply to. The position is in codon
coordinates where the rule applies to amino acid changes and is 1-based inclusive. If
not provided, the end of the gene is inferred.</li>
<li><code>drug</code>: A semi-colon-delimited (<code>;</code>) list of drugs the rule impacts. If the rule
confers susceptibility, use <code>NONE</code> for this column.</li>
</ol>
<p>If there are certain rules you need for your
species-of-interest, <a href="https://github.com/mbhall88/drprg/issues/new/choose">raise an issue</a>, and we can look at implementing it.</p>
<h3 id="example-2"><a class="header" href="#example-2">Example</a></h3>
<p>This is an example of the <em>M. tuberculosis</em> expert rules file used in our paper.</p>
<pre><code class="language-csv">missense,rpoB,426,452,Rifampicin
nonsense,rpoB,426,452,Rifampicin
frameshift,rpoB,1276,1356,Rifampicin
nonsense,katG,,,Isoniazid
frameshift,katG,,,Isoniazid
absence,katG,,,Isoniazid
nonsense,ethA,,,Ethionamide
frameshift,ethA,,,Ethionamide
absence,ethA,,,Ethionamide
nonsense,gid,,,Streptomycin
frameshift,gid,,,Streptomycin
absence,gid,,,Streptomycin
nonsense,pncA,,,Pyrazinamide
frameshift,pncA,,,Pyrazinamide
absence,pncA,,,Pyrazinamide
missense,katG,315,315,Isoniazid
missense,gid,125,125,Streptomycin
missense,rpoB,425,425,Rifampicin
missense,gid,136,136,Streptomycin
</code></pre>
<p>The row</p>
<pre><code class="language-csv">frameshift,pncA,,,Pyrazinamide
</code></pre>
<p>says that a frameshift <em>anywhere</em> within the <em>pncA</em> gene will cause resistance to
Pyrazinamide</p>
<pre><code class="language-csv">nonsense,rpoB,426,452,Rifampicin
frameshift,rpoB,1276,1356,Rifampicin
</code></pre>
<p>these two rules illustrate the context of the start and end coordinates. In the first
row, we say that any nonsense mutation between 426 and 452 in <em>rpoB</em> causes resistance
to Rifampicin. As nonsense mutations only apply to amino acid changes, the coordinates
are in codon-space. Whereas the second row describes a frameshift, which only applies to
nucleotides; therefore, 1276 and 1356 are in bases-space (i.e. the 1276th
nucleotide/base). (As an aside, these two rules both apply to the same region -
the <a href="https://doi.org/10.1016/j.cmi.2016.09.006">RRDR</a>)</p>
<pre><code class="language-csv">missense,katG,315,315,Isoniazid
</code></pre>
<p>describes any missense mutation at position 315 in <em>katG</em> causing isoniazid resistance.</p>

                    </main>

                    <nav class="nav-wrapper" aria-label="Page navigation">
                        <!-- Mobile navigation buttons -->


                        <div style="clear: both"></div>
                    </nav>
                </div>
            </div>

            <nav class="nav-wide-wrapper" aria-label="Page navigation">

            </nav>

        </div>




        <script type="text/javascript">
            window.playground_copyable = true;
        </script>


        <script src="elasticlunr.min.js" type="text/javascript" charset="utf-8"></script>
        <script src="mark.min.js" type="text/javascript" charset="utf-8"></script>
        <script src="searcher.js" type="text/javascript" charset="utf-8"></script>

        <script src="clipboard.min.js" type="text/javascript" charset="utf-8"></script>
        <script src="highlight.js" type="text/javascript" charset="utf-8"></script>
        <script src="book.js" type="text/javascript" charset="utf-8"></script>

        <!-- Custom JS scripts -->

        <script type="text/javascript">
        window.addEventListener('load', function() {
            window.setTimeout(window.print, 100);
        });
        </script>

    </body>
</html>
